File: 백준/Gold/11657. 타임머신/타임머신.py
n, m = map(int, input().split())

edges = []

for _ in range(m):
    s, e, t = map(int, input().split())
    edges.append((s, e, t))


def bellmanford():
    global edges, n, m

    INF = int(1e9)
    times = [INF] * (n + 1)
    times[1] = 0

    for i in range(1, n + 1):
        for j in range(m):
            sour, des, time = edges[j]
            next_time = times[sour] + time

            if times[sour] == INF or times[des] <= next_time:
                continue

            times[des] = next_time

            if i == n:
                return []

    return times[2:]
    


result = bellmanford()

if not result:
    print(-1)
    
else:
    for cost in result:
        print(cost if cost != int(1e9) else -1)

File: 백준/Gold/11657. 타임머신/타임머신.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer tokenizer = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(tokenizer.nextToken());
        int m = Integer.parseInt(tokenizer.nextToken());

        List<int[]> edges = new ArrayList<>();

        for(int i=0 ; i<m ; i++){
            tokenizer = new StringTokenizer(br.readLine());
            int sour = Integer.parseInt(tokenizer.nextToken()); 
            int des = Integer.parseInt(tokenizer.nextToken());
            int weight = Integer.parseInt(tokenizer.nextToken());

            edges.add(new int[]{sour, des, weight});
        }

        long[] costs = bellmanFord(n, m, edges);

        StringBuilder answer = new StringBuilder();
        
        for(int i=2 ; i<costs.length ; i++){

            answer
            .append(costs[i] == Integer.MAX_VALUE ? -1 : costs[i])
            .append("\n");
        }

        System.out.print(answer.length() > 0 ? answer : -1);
    }

    private static long[] bellmanFord(int n, int m, List<int[]> edges){
        long[] costs = new long[n + 1];
        Arrays.fill(costs, Integer.MAX_VALUE);
        costs[1] = 0;

        for(int i=0 ; i<n ; i++){
            for(int[] edge : edges){
                int sour = edge[0], des = edge[1], cost = edge[2];

                if(costs[sour] == Integer.MAX_VALUE || costs[des] <= costs[sour] + cost)
                    continue;
                
                if(i == n - 1)return new long[0];
                costs[des] = costs[sour] + cost;
            }
        }

        return costs;
    }
}
